### ISSUE TITLE
A data-friendly alternative to `Nullable`

### ISSUE THREAD

**Issue by andyferris** [JuliaData/Roadmap.jl#3]
The semantics of `Nullable` have been under heavy discussion over in Julia base at https://github.com/JuliaLang/julia/pull/19034#issuecomment-255247516.

Personally, I had been somewhat confused as to the core problems but @johnmyleswhite makes a very compelling argument which really solidifies (for me) why there has been tension regarding adding features to `Nullable`. I'll copy it here (I hope you don't mind, John).

> There are two almost totally irreconcilable reasons you might care about Julia's `Nullable` type:
> - You are a software engineer and you deal with things that are "null" -- like null pointers and null handles to databases. You absolutely do not want such a value to propagate automatically. Many systems have been rebuilt almost from scratch (Javascript in Flow and PHP in Hack) in part because automatic propagation of nulls is a nightmare for large-scale software design.
> - You are a data analyst and you want to deal with missing values. You want to be able to execute arbitrary expressions against databases that may contain missing values and you want these missing values to be propagated automatically.
> 
> Julia's `Nullable` type was not meant to be optimized for either use case-- it was meant to be a building block for other packages to expand upon. This has come into direct conflict with the community's increasing desire to prevent type piracy. There isn't a simple solution, but it's worth keeping in mind that many of the proposals in this thread aren't Pareto improvements for the Julia community: allowing expressions like `1 + Nullable(1)` is likely to do harm to software engineers in order to benefit data analysts.

To me, if we are going to make a way forward, we should really begin developing a data-friendly alternative to `Nullable` and leave `Base.Nullable` for the software engineers. I think this will allow for much more rapid progress. My proposal would be to create a type which behaves semantically as close as possible to `Union{T, NA}` while remaining type-stable, where `NA` is the missing value type in the _DataArrays.jl_ package which behaves somewhat like `NaN` does for `Float64`.

However, this needs to be a discussion for the community, and it doesn't necessarily have to involve `Base` Julia at all. I'll ping a bunch of people and see what happens.

@johnmyleswhite @JuliaData (not sure if that works so I'll just add everyone @ararslan @davidagold @dmbates @kleinschmidt @nalimilan @quinnj @richardreeve @scidom @shashi @simonbyrne )

**Comment by ararslan** [comment]
I was actually thinking the same thing going through the base thread, that it may be worthwhile to pursue an entirely separate type for statistical missingness. Thanks for getting this discussion started, Andy, much appreciated!

As we think about this, we should also take a hard look back at DataArrays, which was the initial attempt to do exactly this. What did and did not work well?

I'll also cc @JeffBezanson and @StefanKarpinski here.

**Comment by andyferris** [comment]
> What did and did not work well?

I'll start with the obvious and say type stability was a "slight" issue...

**Comment by ararslan** [comment]
We should also be careful not to negate the massive amount of work many of the JuliaStats folks have put into making `Nullable`s work in a statistical context.

**Comment by davidanthoff** [comment]
I like this idea *a lot*. In fact, the more I think about it, I think I'll just use that approach for now in Query.jl. There is a super simple way for me to make Query.jl still work with sources and sinks that are based on ``Nullable``, but inside I would use a ``NAable`` type...

**Comment by andyferris** [comment]
Sure. There's no reason we can't define `convert(::Type{NAable}, ::Nullable)` (or whatever we call this new type).

On that note, the obvious bikeshedding - what is a good name for a stats-y nullable type?

 * `Optional{T}`
 * `Maybe{T}`
 * `NAable{T}`
 * `NA{T}`
 * `Missing{T}`
 * `?{T}` (yes this actually works)

**Comment by davidanthoff** [comment]
Given that this type is meant to squarely target data scientists, it would be nice to stay as close as possible to names/concepts that folks in that area know. I think that speaks for something with ``NA`` in it. If we were to go with ``NAable``, we could define ``const NA = NAable{Union{}}``, and then one could even write things like ``a == NA`` instead of ``isna(a)``, which I think would make this even simpler to use.

**Comment by ararslan** [comment]
I dislike `NAable` (and I also dislike the precedent in R of `NA` denoting missing). `Optional` and `Maybe` I think are more computer science-y than would be familiar for most statistically minded folks. Of @andyferris' suggestions, 👍 to `Missing{T}`.

**Comment by andyferris** [comment]
> we could define const `NA = NAable{Union{}}`

I'm surprised `Nullable{Union{}}()` actually constructs... I have to say that is kind-of cool. I do support having a generic null *value* that can be `isequal` or maybe `==` to others.

**Comment by ararslan** [comment]
> one could even write things like `a == NA` instead of `isna(a)`, which I think would make this even simpler to use.

I think a behavior like this could be nice, though in general I think I prefer using `is*` functions for things like this. But if we do go with `NA`, that has the potentially to be _very_ confusing for people coming here from R, since `anything == NA` in R is `NA`.

**Comment by davidanthoff** [comment]
``Missing`` as the type seems odd to me, after all sometimes the value is *not* missing :) Maybe ``Missable`` for the type, and then ``const missing = Missable{Union{}}()``? On the flipside, ``missing`` is long, ``NA`` is much nicer on that score.

>But if we do go with NA, that has the potentially to be very confusing for people coming here from R, since anything == NA in R is NA.

That is a good point...

**Comment by ararslan** [comment]
"Missable" is a word but it means something unrelated.

**Comment by andyferris** [comment]
Looking at some other languages (from [wikipedia](https://en.wikipedia.org/wiki/Option_type)), there are only two common classes of naming:

 * Many languages use a typename like `Optional` and have a null install `none`, `nothing` or `null` and many seem use the word `some` to construct or pattern match a non-null value.
 * Haskell (and idris): Type `Maybe`, null instance `Nothing` and other instance `Just x`.

`Optional` is overwhelmingly most common on wikipedia's list. Doesn't mean we can't go our own direction.

> "Missable" is a word but it means something unrelated.

Yes I don't think `Missable` is quite correct.

**Comment by nalimilan** [comment]
I've contemplated creating an alternative nullable-like type a few times, but honestly I'm not sure it would solve all problems.

1) The argument that we shouldn't repeat the same mistakes as for vectorized versions of functions is fundamentally right as regards the general case: we cannot reimplement all functions manually to have lifting semantics. So we need a generic system to lift a function at the call site. The framework at https://github.com/JuliaLang/julia/pull/18758 fills most needs, and I think some syntactic sugar like `f?(x)` or `f.?(x)` which @JeffBezanson mentioned would be nice. Since that's explicit, it isn't an issue for people who don't want null propagation.

2) There's actually an argument for *not* using a custom type: since `Nullable` is a standard type, we can imagine introducing syntactic sugar for it (as in many other languages), like `Int? -> Nullable{Int}`, `1? -> Nullable(1)`, etc. We can't have this with a custom type. Same if you would like all functions to automatically lift by default: this would be possible only via a change in Base.

3) So the main (or maybe only) conflict regards lifting operators and mixing nullables and scalars. People who don't want nulls to propagate don't want standard operators to have lifting semantics, and it might indeed be hard to satisfy both conceptions of `Nullable`. I don't have a good solution to that, except maybe to suggest that we use `Nullable` for missing data semantics, and create a new `Option` type in Base with a strict non-propagation behavior. Both types would behave the same in most contexts, except with regard to operators. Not sure it's worth it (but the definition of `Nullable` is only a few lines, so it's not a big cost either).

4) I don't think we even agree about the semantics `==` should have on nullables: should it return, `Bool` or `Nullable{Bool}`? should `NULL == NULL` return `true` or `false`? I don't even know myself what solution I prefer. So adding a separate type wouldn't solve that debate either (or at least not yet).

**Comment by andyferris** [comment]
I totally agree there will be (major!) difficulties, @nalimilan. And there is a major issue of semantics that need to be discussed

In my opinion, I see two self-consistent, useful, sensible semantics for a nullable type. I think we should implement both.

 * A container which has zero or one elements. All unpacking is done by the user. It is safe for purposes where automatic constant propagation is better to avoid. This will be `Base.Nullable` (if only for the fact that that is how Jeff thinks `Nullable` should behave, and its more-or-less how things stand now).

 * A new type such as `NA{T}` which *aims* to be equivalent to `Union{NULL,T}` in a semantic sense. I think since we have `isequal` to solve `NaN` difficulties, we can have `NULL != NULL` but `isequal(NULL,NULL) = true`. This makes `NULL` and `NaN` share similar properties. We will have problems making it work for *all* functions automatically, but we can aim to solve 95% of common cases.

In the future, as ideas progress and Julia evolves, I'm betting we'll find nicer ways of working with the former (e.g. using `.` broadcast syntax or new syntax with `?`). 

Even if the latter type is implemented outside of `Base` (for now) I'm also betting working with it will get easier. E.g. `NA{T}` could inherit traits from `T` in a future version of Julia that easily allows traits for dispatch. Or if we solve https://github.com/JuliaLang/julia/issues/14919, then maybe we can define things like 

```julia
(f::Function)(x::Nullable) = hasvalue(x) ? f(x.value) : NA{return_type(f, Tuple{eltype(x)})}()
```

There could be a million other helpful things we can't even imagine yet. Like @davidanthoff said quite well [here](https://github.com/JuliaLang/julia/pull/19034#issuecomment-255010621), we shouldn't give up on trying just because its a little disgusting or difficult or imperfect on our first try.

**Comment by nalimilan** [comment]
I don't see anything in your proposal which says what using a new type would allow us doing *right now*. How would it help getting type-stable nullable arrays/data frames ready for a release in the next few weeks (or even by the time Julia 1.0 is released)?

The `(f::Function)(x::Nullable)` approach is certainly appealing, but it would only work for one-argument functions. It's hard to design a similar mechanism which would work for any number of arguments (both nullable and non-nullable) without introducing lots of ambiguities with any function which accepts nullables. I think we would need changes inside Julia to get this to work. So, yes, maybe at that point we could introduce a separate type which would automatically lift arguments, but that would be the final step after a full design has been implemented inside Base.

(See discussion a https://github.com/JuliaStats/NullableArrays.jl/pull/85 about `==` semantics. In general I encourage you to read all issues linked in the various related PRs as well as in https://github.com/JuliaStats/DataFrames.jl/pull/1008. There are been quite a lot of design discussions all over the years.)

**Comment by davidagold** [comment]
Call overloading on functions doesn't seem to be permitted:

```
julia> (f::F){F<:Function}(x) = f(x)
ERROR: function type in method definition is not a type
```

But if it were, writing `(f::F){F<:Function}(xs...) = lift(f, xs...)` would immediately handle argument signatures of all lengths and "mixedness".

**Comment by davidanthoff** [comment]
> I don't see anything in your proposal which says what using a new type would allow us doing right now. 

I'm converting Query.jl to use a new type ``NAable`` now, and it pretty much solves all the problems I had in that area. The main benefit is that I can now implement the white-list approach without having to a) convince folks that it should be in base (seems hopeless at at this point) and b) without type piracy. I should also add that I actually don't have any open question in terms of what the semantics of these lifted functions should be within the context of Query.jl (things like ``==`` etc.), so having my own type allows me to just implement those semantics.

This approach works well for Query.jl, even if the rest of the data universe doesn't adopt the ``NAable`` type. Right now I simply convert anything ``Nullable`` into a ``NAable`` at the start of any query, and then the other way round when someone collects it into a data structure where one would expect a ``Nullable``.

Obviously this doesn't help with the old API stuff like ``df[:a] = df[:b] + 2`` etc., but at least for Query.jl I think that pretty much solves the problem, and it should also make Query.jl pretty usable for a release of ``DataFrames`` that uses ``NullableArrays``.

I'd be happy to a) rename the type to something better if there is a consensus and b) move all of that into its own package eventually. But for now I just want to move forward with Query.jl, so I'll just have it there for now.

**Comment by quinnj** [comment]
It's certainly great to have someone charge forward on this approach. One of my biggest concerns was just finding someone to implement and maintain the code. I think it'll be great to see how things unfold with it's usage and it will provide some good chances to learn as we go (separate from Base and with more flexibility to change/update things).

**Comment by andyferris** [comment]
Thanks @davidanthoff for going forward with this. I was going to tell @nalimilan that unfortunately I quite probably wouldn't have time to create an entirely new and complete system in a few weeks, but it seems that David will. :)

Though, I would recommending splitting off `NAable` to a package ASAP. This will allow you to journal your design decisions cleanly and for feedback/feature requests from the community and for others to contribute and make PRs - the administration effort should pay off for you in terms of contributed ideas and code (and I find users are really good at finding bugs). I know you want to move quickly with *Query.jl* but I hope this wouldn't get in your way (if the package changes API or even name while it is quite young, you won't be annoying anyone too much).

**Comment by andyferris** [comment]
I've been reading (and re-reading) about earlier discusions surrounding `==` as Milan suggested, and I see David and Milan (and many others) have been over all of this before. Though, I still firmly believe that users will sometimes want `NULL == NULL` to be `false` and sometimes they will want it to be `true`. Milan wrote

> I've done a small review of choices made by other languages, and indeed they don't seem to bother about the inconsistency between NaN and NULL. All languages I surveyed which support == for the equivalent of NULL consider two NULL values as equal. In Python, None == None -> True (though using is is recommended). In Ruby and Swift, nil == nil -> true. In C++, C#, Java, JavaScript and Kotlin, null == null -> true.
> 
> Only SQL, R and SAS return the equivalent of Nullable{Bool}, and there NULL == NULL -> NULL.

However I'm not sure (I could be wrong) those languages had a well thought out system of `==` and `isequal`, and `<` and `isless`, like Julia has had since the beginning. I think a lot of thought went into those features from the Julia designers, and I really appreciate it. And despite the fact that all those other languages chose something different in the past, I don't think that means we have to do the same (caveat: I realize *Query.jl* aims to be close to LINQ syntax). I really think we should follow `NaN` semantics, and from my experience learning Julia I discovered the difference between `==` and `isequal` relatively early on so I don't feel this will be a huge burden to new users. After all, the idea is simpler to explain than three valued logic.

Another thing - I saw a comment equating hashing and `isequal` by Stefan. It would be nice to ensure `NAable` hashes in a deterministic way if it is `NULL` (maybe a clever inner constructor could zero out the `T` bits, or we overload `hash`).

**Comment by nalimilan** [comment]
> Though, I still firmly believe that users will sometimes want NULL == NULL to be false and sometimes they will want it to be true.

And sometimes they will want it to be `NULL`, which is the hardest decision to make.

> Another thing - I saw a comment equating hashing and isequal by Stefan. It would be nice to ensure NAable hashes in a deterministic way if it is NULL (maybe a clever inner constructor could zero out the T bits, or we overload hash).

That's easy: don't hash the value of null.

**Comment by davidanthoff** [comment]
Here we go: https://github.com/davidanthoff/NAables.jl

I also have a branch in Query.jl that uses ``NAable`` instead of ``Nullable`` everywhere: davidanthoff/Query.jl#70. It works without problems with any source or sink that uses ``Nullable``, so from an interoperability point of view this doesn't really change anything.

Right now all the code in NAables.jl is also in Query.jl, but that is just until we've decide on a final name for NAables.jl and METADATA registration.

I'm still somewhat nervous about the Query.jl switch, so I'll probably wait a little longer before I merge this. But right now it solves a huge issue, namely I can get rid of all type piracy, and I don't really see any downside...

**Comment by ExpandingMan** [comment]
I think the answer to the question of the behavior of `Nullable(a) == Nullable(b)` is simple: it's `false`.

This is quite simply because `false == (NaN == NaN)`.  It's always been this way, and it's always worked fine, in fact that's why we have `isnan`.  I think the behavior of `NaN` is a very sensible guide to how to make a data nullable type behave.

**Comment by nalimilan** [comment]
No, it's definitely not that simple (read discussions I liked to above). For example, R uses a completely different approach. There are several legitimate strategies, and it's hard to know which one is best. There are also subsidiary questions, like do we want `null != null` to return `true` (like for `NaN`) or `false` (for consistency with `==`)?

**Comment by davidagold** [comment]
Having `Nullable{T}() != Nullable{T}(x)` seems like a great way to subtly invalidate an analysis, *especially* if the rest of your data manipulation framework propagates null values.

**Comment by ExpandingMan** [comment]
I don't know, I've worked with NaN's for a long time.  In fact, before my current job I was a physicist, and since I dealt almost exclusively with doubles, `NaN` was pretty much the only type of null value I ever used, and I never, ever had any reason to complain about it.  In fact, if I were still in that position and didn't have to use annoying data types like `String` and `DateTime`, I'd probably only ever use NaN now and would be completely disinterested in this whole discussion.  Furthermore `NaN != NaN` was a nice easy way of checking for them.  

I guess I have to concede that not everyone shares this experience, but I would like to make one further point: `NaN` is an ultra-standard thing that's been around for 40 years in all sorts of code used for all sorts of purposes.  R (and `pandas` for that matter, as they do a really abysmal job of dealing with this particular issue, but I guess it is kind of the canonical example of what's wrong with python), on the other hand, is a niche thing.

**Comment by nalimilan** [comment]
If you think R and Pandas are a niche thing, would you also qualify SQL as such (cf. [Wikipedia on NULLs in SQL](https://en.wikipedia.org/wiki/Null_(SQL)#Comparisons_with_NULL_and_the_three-valued_logic_.283VL.29))? Then what's left in the realm of data tables management?

**Comment by ExpandingMan** [comment]
Sorry, I was probably a bit overzealous in my language.  I was just trying to make the point that floating point NaN is a very standard, universally understood thing in a way that certain conventions in R and pandas are not.  I'm still arguing for using NaN as the paradigm, but completely understand that there are other, perfectly legitimate ways of doing things.

**Comment by StefanKarpinski** [comment]
Based on a Slack conversation, this is what I feel needs to be done for 1.0:

1. Remove `Nullable{T}` from Base.
2. Aad `Some{T}` to Base
3. Use `Union{Some{T}, N}` and `Union{T, N}` variously in place of `Nullable{T}` (depending on whether `N <: T` or not), where `N` is either `Void` or some other engineer’s null type.
4. Decide what `N` above is: it could be `Void` with instance `nothing` or some new `Null` type with instance `null` (presumably).

Things that don't need to be done for 1.0:

* Add a data scientist’s null to Base. This may never need to happen: there are different semantics of missing data and people may want to use different ones or even many at the same time. The bits-type union optimization work means that user-defined null types are automatically efficient.

* Give a meaning to `T?`. Even though we want to use this syntax for some kind of missing data, now that we've freed up the syntax, we can reclaim it at any point in the 1.x release series. We may end up wanting this syntax for the built-in engineer's null, for some "standard" data scientist's null, or let it be user-definable so that different people can decide for themselves what they want it to mean.

**Comment by quinnj** [comment]
I respectfully disagree on things that don't need to happen for 1.0. How long are we going to keep punting on the missing data story? What are the valid concerns against using Union{T, Null} as _one_ standard way to do missing data, with nice syntax in Base? Having been there, I know this exact topic was discussed extensively at JuliaCon 2017 with all parties coming to the conclusion that it was a good way forward and would help solidify Julia's "data" story once and for all. I also don't remember a single concern not having been resolved to everyone's satisfaction. I just don't think we should punt on this issue any more; we've now iterated for what, 4-5 years from DataArrays to NullableArrays to Nullable in Base and now we've settled on Union{T, Null} as the most basic representation, which also enables other missing data stories to build upon (i.e. an explicit unwrapping software engineer's nullable as `Some{T}?`, a container-based missingness story like DataValues purports, etc.).

I just feel there's much more to gain by making the commitment for 1.0 than having to release Julia to the world of data processing/analytics/machine learning/data science with a little asterisk that "oh sorry, we don't have any official/built-in support for missing data yet; maybe check back in a couple years for 2.0".

For real though, if people have real valid concerns about moving the ~100 LOC of Nulls.jl into Base and providing the `T?` syntax to mean `Union{T, Null}`, let's get it all out in the open right now, mainly because many of us feel like we already hashed this exact discussion out this summer and came away with everyone ok with that gameplan.

**Comment by rofinn** [comment]
> What are the valid concerns against using Union{T, Null} as one standard way to do missing data, with nice syntax in Base?

@quinnj For those of us that weren't at JuliaCon 2017 do you remember if/how the software dev and stats requirements were unified?

My understanding so far is that there are currently 2 use cases that would have different behaviour.

1. `Option{T}` (software development) - Represents `Union{T, nothing}` and simply provides a safe type stable way of representing missing values, but nothing else.
2. `NA{T}` / `T?` (statistics) - Represents `Union{T, NAType}` and has specific rules around promotion and propagation.

NOTE: I'm not super opinionated about the names, but I'd like to have very different language if we need to deal with different concepts.

I could also see this pattern being useful outside of missing values (e.g., `Result{T, E}`
 in [ResultTypes.jl](https://github.com/iamed2/ResultTypes.jl) could become `Union{T, E<:Exception}`).

**Comment by davidanthoff** [comment]
I agree with @StefanKarpinski that we should try to keep the data science story out of base as much as possible for 1.0. I don't think things are settled enough to move it in, plus I think there is real value in being able to iterate quickly on that stuff and not be tied to the julia release cycles.

> I also don't remember a single concern not having been resolved to everyone's satisfaction. 

That is not an accurate description of where we are. We still haven't resolved how ``Union{T,Null}`` could be used in Query.jl. It was also not resolved at juliacon. Various folks had ideas, but they were partly conflicting and in the end inconclusive. I would be super thrilled if there is a good solution, but so far I haven't seen it. I should also say that I thought about this all summer and tried various things, but couldn't find a solution. So I'm not super optimistic on this front :)

>  which also enables other missing data stories to build upon (i.e. an explicit unwrapping software engineer's nullable as Some{T}?, a container-based missingness story like DataValues purports, etc.).

I don't understand the part about ``DataValues`` here.

Here is where we are in my mind: we are pretty sure that ``Union{T,Null}`` will work well for the existing ``DataFrame`` interface, but we haven't found a solution for Query.jl. We know that ``DataValue`` works really well for Query.jl, but we don't know whether one could have a ``DataValueTable`` (i.e. a table type that is based on ``DataValue`` for missing data) that is easy to use. In my mind we should just continue to explore both avenues for missing data (``Union{T,Null}`` and ``DataValue{T}``) and see what happens. Hopefully we'll find a way to make one of those work for everything, and then we can declare victory. But from my point of view it is not clear at this point which of those is the more promising strategy, so hedging our bets to some degree seems a good idea to me.

**Comment by nalimilan** [comment]
From recent discussions, I think the plan is becoming clearer:
- Use `Union{Some{T}, Void}` as a replacement for `Nullable{T}` in Base; `Union{T, Void}` will also be used in cases where the wrapper isn't needed (i.e. `nothing` cannot happen except when there is no value).
- Keep `null`/`Null` in Nulls.jl so that it can evolve independently from Base and be suited to the needs of the data ecosystem without interfering with requirements from other fields.
- Leave the `?T` syntax free for use by packages. Ideally, since two packages cannot use it at the same time, it won't be claimed by too many packages, at least not by packages which are likely to be used together (else the user will have to choose).

**Comment by rofinn** [comment]
@nalimilan Would there be any benefit to using a different missing sentinel value though (e.g., `Union{Somt{T}, Void}` vs `Union{T, NAType}`)?

**Comment by nalimilan** [comment]
I'm not sure, we could discuss that. The idea was to keep `Null` in Nulls.jl, so that leaves either 1) use `Void` or 2) introduce a new type in Base. The latter option means there would be two kinds of "nothingness" in Base (in addition to things like `#undef` and the empty tuple/array), plus one (or even two with `DataValue`) in packages. I'd say the increased complexity isn't worth it, as users will likely be confused (cf. https://github.com/JuliaLang/julia/issues/23721#issuecomment-331869022)).

**Comment by amellnik** [comment]
If nothing else, using `Null` makes it explicit how missing values will be treated (as distinct from the various uses for `Void`, etc).  It also makes it easy to see when packages are "bought in" to the Nulls.jl approach verses `Nullable{T}`,  `DataArrays`, etc.

**Comment by nalimilan** [comment]
As I said, the goal was to keep `Null` in Nulls.jl, so it couldn't be used in Base. Hence the idea of using `Void` instead.

**Comment by richardreeve** [comment]
So, I appreciate I'm coming into this a bit late, and apologies for the length of this, but I'm feeling a bit confused here. As a (former) mathematician and former software developer now using Julia in the life sciences (for I guess what you would broadly call data science), I can see the need for these two distinct things - optional arguments and missing data - and the broad outline of what's being discussed is fine in that respect. However, I'm completely with @quinnj that we need a solution for the data science issue and we need it in 1.0. The arguments for me break down to:

- I would say that _at a minimum_ 50% of the people in the life sciences departments in my university, like me, have to deal with missing data, and aside from me, and a handful of other lunatics, they all use R (or in the case of older people sometimes SAS) to do this (okay, some use python too!).
  - I don't think my university is remotely atypical. I haven't met a researcher for years in the life sciences who doesn't use R (or occasionally python) or know they should be using it, so suggestions from @ExpandingMan that R is niche are misplaced, unless the life sciences as a whole are niche.
- The ability to handle missing data is critical to any chance of adoption of Julia in the life sciences, and automatic propagation of missing data is critical to that (and I assume in data science in general).
  - Any software that thinks that in a missing-data-context `NA == NA` or `NA != NA` or `someval == NA` or `someval != NA` is **wrong**. This isn't a discussion as far as I'm aware. It's just wrong. the only possible answer to avoid catastrophic errors in analysis and interpretation is that any comparison between NA and another thing of any type returns NA, because the missingness is propagated in that sense as in all others. It goes without saying that if a datapoint is missing, we don't know if it is equal to another datapoint, whether or not that datapoint is missing itself, so the answer is missing.

Having said that I look at the proposed solutions and I find that:

- `DataValue{Int}() == DataValue{Int}()`, so this seems to be without value as an approach.
- `null == null`, so I don't understand how using `Union{T, Null}` will help us.
- `NaN != NaN`, so that doesn't help either.
- on the other hand `Nullable{Int}() == Nullable{Int}()` throws a null exception, so at least that's an improvement.

I'm clearly missing something here, hence my confusion. To add to it though:
- the plan is to remove `Nullable{T}` from Base Julia - I don't object to this if there's a better replacement, but...
- there is no plan to put any solution into Base Julia according to @StefanKarpinski in any kind of a hurry.

So, as I see it, there is no prospect for the huge number of people who work with missing data will adopt Julia, and there is surprisingly little concern about this from some of the Julia core team. Understandable concern from @davidanthoff that he can't see how to work some of the proposed solutions into `Query.jl` is fatally undermined in my mind (sorry) by the fact that if `Query.jl` is currently based on `DataValue{T}` it handles missingness incorrectly, so is irrelevant to the debate.

I would just plead with everyone involved in this that we really, really do need a solution to missing data in data science where missingness propagates correctly before 1.0, whether or not it completely breaks packages like `Query.jl` (sorry again!).

I personally really like the idea of the idea of `Union{T, Null}` as a solution if null is changed to respond correctly to comparisons, but that is completely irrelevant - anything that works and is in Base is fine by me.

PS The idea of using `T?` is fantastic - fundamentally really cool syntax for whatever the final solution is - but absolutely also should be in Base. @nalimilan's idea that it can be left out and then mean different things in different packages would be a catastrophe (in my head). Sorry again for the length!

**Comment by nalimilan** [comment]
@richardreeve I'm glad to hear one voice in support for `null == null` returning `null`, since after lots of discussions this is precisely what we've decided to do at https://github.com/JuliaData/Nulls.jl/pull/33 (not included in a release yet). I'll let you debate (elsewhere) with @davidanthoff about whether `DataValue` should work the same.

Regarding your other concern about `null` not being included in Base, please don't take it as an indication of a lack of consideration/interest by the core developers. Precisely, that's because missing values support is so central to the success of a scientific language like Julia that nobody wants to set a design in stone in Julia 1.0 and run the risk to get it wrong, and regret it later. Being in Base or not makes (almost) no difference for Julia types, in particular in terms of performance. For example, `DataFrame` lives in an external package, yet it's a central type in the data ecosystem.

What matters is that this ecosystem agrees on a common representation missing values, and that's almost completely orthogonal to where `Null` is defined. Yes, it would be confusing and annoying if the `T?` syntax is claimed by different competing packages in the same field, but that's a matter of coordination and social engineering, rather than a technical question of where to define the type. If  we can't agree on a common representation for nulls, the ecosystem is going to be inconsistent anyway.

**Comment by ararslan** [comment]
I agree with you 100%, @richardreeve. I also think not having `null` in Base with `T?` centrally claimed (and _not_ overloadable) in 1.0 will be an enormous mistake. I think it's a very real possibility that it would cause people to overlook Julia for data analysis in 1.0.

**Comment by richardreeve** [comment]
@nalimilan I am delighted to hear that `null==null` now returns `null` on master in JuliaData/Nulls.jl! Fantastic. I think this is a great step forward. I can sympathise with people concerned about the 3VL-related issues that arise as a result of this, but not enough to stop believing the simple truth that we don't know if two missing values are equal to each other or not.

Now that this has happened, I would advocate even more strongly for `null` and `T?` to be merged into Base. While technically and from a coding perspective I completely understand that it makes no or little difference whether it's in Base or not, in reality I think it makes an enormous difference.

I've persuaded one of my PhD students to do her entire PhD in Julia because it would be computationally intractable in R (or some other scripting language like python or ruby) and it would otherwise be necessary to do it in C++ or some other compiled language. However, I would not currently dream of trying to move my less quantitative students over to Julia, because it does not currently have core support for some of the fundamental things that are necessary to our work. I can live happily without some kind of data frame in Base, because there are so many ways of handling this kind of data depending on size and complexity, and I'm really happy to see so much exciting development work going into this - not least the really cool stuff we saw at JuliaCon on `JuliaComputing/JuliaDB.jl` (which I really must find time to play with!), but _conceptually there simply aren't that many ways of handling missing data_.

Not being able to agree on just one of that handful of options means that different packages will inevitably handle it differently (see `davidanthoff/Query.jl` vs `JuliaData/DataFrames.jl`, for instance) not just in detail but **in the really important sense that you get completely different answers in different packages when you do the same thing**! This tells me in so many words that we don't really consider missing data, and therefore data science, to be an important use case for the language. If that were true I shouldn't be using Julia, but I obviously don't believe it is.

However it does mean that now and for the foreseeable future I can't sit down with my students and say, **this** is how it works, we have this cool syntax for it (`T?`, natch) that is crystal clear in its meaning and incredibly simple to use and always **always** means the same thing because it can't be overloaded (great idea, @ararslan). @StefanKarpinski **please** commit to putting it into 1.0.

**Comment by andyferris** [comment]
It seems to me that `null == null -> null` is semantically correct, with `isequal(null, null) -> true` also making perfect sense. This seems consistent with `NaN` and I would go as far to say that the IEEE standard was cornered on `NaN != NaN` because of e.g. static typing and performance concerns.

@richardreeve I do agree that I'd strongly prefer a good story on nulls in `Base` and the sooner the better - however have you considered that *new features* can be added in v1.1, etc? (I.e. we don't have to wait until v2.0). IIRC the release cadence has typically been 9-12 months - significantly shorter than a PhD, and I assume typing something like `using Nulls` in the meantime wouldn't be such a drag, and (It's my hope that the ecosystem will come to consensus prior to it appearing in `Base`, though if not that's also a good way of enforcing consensus).

However, surely its clear that syntax like `T?` should be claimed by `Base`? (I'm thinking of the historical `~` here... leaving this special syntax undefined didn't work out as well as planned IMO). There's also the consideration that there is probably code in `Base` that goes `if a == b` which we need to "upgrade" to be `null` proof.

**Comment by ExpandingMan** [comment]
Concerning the issue of whether it is necessary to put `null` in `Base`, I'd just like to point out that pandas to this day still does not have a comprehensive way of dealing with missing data.  In fact, over there the situation is particularly absurd since the type of missing value depends on the context (i.e. what the data type of the underlying numpy array for a particular column is) which python to some extent, infuriatingly tries to hide or otherwise obfuscate from the user.  Furthermore, whether one is supposed to use `None`, `nan` or `NaT` depends on the data type and I can tell you from experience that people often do not respect this distinction and totally screw the pooch on performance by using `None` everywhere.

Point is, we are already in a much better situation than pandas.  Considering the massive popularity of pandas, it's hard to understand why it would be necessary to have `null` in `Base`.  Apparently most users just don't care that much.  That doesn't mean we *shouldn't*, of course we *should*, but it's hard to imagine too many people saying "Oh, I have to do `using Nulls`?  Blasphemy, I'll never use it!".

Having said all that, I don't necessarily have any objections to putting `Nulls` in `Base`, except to say that we had damn well better get our shit together before 1.0 if that's going to be the case.  It's a little hard for me to imagine this turning into a huge issue that affects the popularity of the language however.  If the Julia userbase really takes off, the conversation will probably seem rather arcane to a sizable fraction of the people coming in.

**Comment by StefanKarpinski** [comment]
This debate is a somewhat surreal experience. Everyone seems to agree that their particular notion of how missing data should work NEEDS to be in the base language, yet there is total disagreement on exactly how such a type should work or even what it should be called. At the same time I have yet to see a _single_ technical reason why a missing data type would need to be defined in the base language in order to do its job well. We've already added the general ability to define and use null-like types efficiently just by using unions. We've also freed up the `T?` syntax so that it can be made overloadable in 1.0 and defined by users and/or packages.

**Comment by StefanKarpinski** [comment]
> However, surely its clear that syntax like `T?` should be claimed by `Base`? (I'm thinking of the historical `~` here... leaving this special syntax undefined didn't work out as well as planned IMO). 

The problem with `~` was not that it was overloadable, but that it was a macro – that was just such a weird quirk of the operator that it ended up being regrettable almost as soon as we introduced it.

**Comment by ararslan** [comment]
Letting `?` be overloadable sounds like complete chaos.

**Comment by richardreeve** [comment]
Your point about the technical lack of need for this to be in Base in 1.0 may well be right, I'm not the right judge, although I think that overloading `T?` to mean different things will be a catastrophe that means that different people get different results by doing the same thing in different environments. This alone will be awful, as @ExpandingMan's example of what apparently happens in python amply demonstrates [as an aside, I don't know a single person who uses pandas, and I personally know hundreds who use R].

But I wasn't arguing the technical issue, but the psychological one. Saying you don't care if we devolve into what sounds like python's complete chaos is a statement that people in the life sciences shouldn't use Julia.

**Comment by StefanKarpinski** [comment]
I never said that I don't care, so please don't put words in my mouth. I'll quote @JeffBezanson here: 

> If data folks want `Null` in Base, they have to agree on how it should work. Can't have it both ways.

Since there is no single coherent story, it's not even clear what we would put in Base. Moreover, I would recommend that the name of the missing value type should suggest its semantics. The word "null" is lousy since it suggests no semantics at all. There are at least three kinds of missing data:

1. There is no value – `true | nothing = method error`
2. There is a value but I don't know what it is – `true | unknown = true`
3. There could be a value but something is wrong or invalid – `true | invalid = invalid`

The first kind is going into Base because it's needed in Base APIs and we're probably going to use the existing `nothing` value for it. But it's not the most useful or popular kind of "null" for data.

**Comment by ExpandingMan** [comment]
Part of the point of my exposition of python's missing data woes is that this topic is probably not quite as important as we are making it out to be here.  The academic scientific community largely won't care because their data is floats and they use `NaN` (this was my experience), and the industrial data science community partially won't care because chaos is the status quo.  Of course we should not have chaos in Julia, but as long as we make a reasonable choice and stick with it I think we'll be fine.  From where I stand, what is currently in Nulls.jl staying where it is but being included by everything in the data ecosystem is perfectly satisfactory (provided the promised performance improvements for `Union` types pan out).  For me that's sort of the end of the story, I don't need to spend any more time stressing about it, and I'm not sure why others would have to either.  Right now a far bigger concern is making sure I can use Julia in an ecosystem that involves lots of python, spark and SQL usage, this seems like a higher order correction.

**Comment by quinnj** [comment]
> yet there is total disagreement on exactly how such a type should work or even what it should be called

> Since there is no single coherent story

These statements are too strong and are probably where some of the frustration is stemming from in this discussion. As far as I know from this and other related discussions, there is only a single dissenter (@davidanthoff) on how a missing value type should behave, be defined, or be called. Everyone else who has chimed in has agreed that `true | invalid = invalid`, which is the behavior on Nulls.jl master.

For me, I certainly understand the hesitance to consider including another "standard library" package at this point before a 1.0 release, but I also firmly believe that this is a topic that has _already been iterated on for years_ in the julia ecosystem, with all design & iteration leading to current Nulls.jl. I also firmly believe that it would be a godsend for the data ecosystem to have a "blessed" representation of missingness, that packages could rely on and build against. It would go a long ways to making the "marketing event" of 1.0 more complete as a swath of data-related packages could also be released using a coherent strategy for missingness. I think Julia stands to "miss out" on more than it gains by leaving them out for 1.0.

**Comment by StefanKarpinski** [comment]
I should point out that any data-frames-like data analysis tooling package should absolutely not make assumptions about what kind of nulls are used. This was one of the major mistakes of the original DataFrames – the `DataFrame` type should be completely agnostic to how columns are represented and the whole business of DataArrays lying about their element type is a disaster (i.e. claiming that the eltype of `DataVector{T}` is `T` instead of `Union{T, NAtype}`). If you want a data column that can hold integers or missing values, its type should be `Union{Int, Missing}`. This means that the user could decide exactly what null semantics they need – it would not be forced on them by the ecosystem. In fact, even if we agree on a single null type for data analysis, the tooling should still be diligent to not assume that since it will result in better factored software to not do so, and it will leave the window open for future innovation in this area. Many people also don't want to have the null support forced on them – it should be possible to have an `Int` typed column that cannot hold any kind of null value. I think we're all on the same page here, but it bears stating explicitly.

**Comment by mlhetland** [comment]
@quinnj From what I can see, the definition is `(|)(b::Bool, a::Null) = ifelse(b, true, null)`? I.e., it 'ould return `true`, not `null` (or, under the renaming, `invalid`)? I guess I'm just misunderstanding :-)

**Comment by quinnj** [comment]
@mlhetland, no, sorry, it was my mistake of saying Nulls.jl follows @StefanKarpinski's description of `invalid`. I confused it with the earlier discussion around `true == null = null`, and failed to look closely enough to see it was in fact `true | null = true`.

**Comment by richardreeve** [comment]
@quinnj's "marketing event" is exactly what I'm talking about - I think this would be a real declaration that the language is intended for use by life scientists, and I could personally use this to good effect tomorrow to argue that we ought to be thinking of a (long-term!) strategy of moving our department's research and ultimately our teaching over to a real programming language.

I also believe that @ExpandingMan is completely wrong both about NaN-like behaviour being okay and about people not caring about inconsistencies. Missing data is _everywhere_ in the life sciences, and behaves exactly how @quinnj and I have described. It's also often not floating point - factors are a big issue in particular. I've personally stopped using R in anger completely and replaced it with Julia, but I will never try to teach a language with contradictory and inconsistent syntax to my students because it's more trouble than it's worth.

I'm not trying to put words in anyone's mouth (sorry, @StefanKarpinski!), I'm just telling you what I believe to be true. As everyone acknowledges, this is a really important thing to get right, but we know how to do it now, and so we should just bite the bullet and do it.

PS None of this requires `DataFrame` to behave in any particular way, as I said above, and it would be a bad thing if it did, as would lying about element types. All it requires is to define in Base some type that behaves like `Null` in `JuliaData/Nulls.jl` master - I couldn't care less what it's called - and then define a non-overloadable `?` postfix operator such that `T?` is just `Union{T, Null}`. Then we have a language that has built-in support for missing data.

**Comment by rofinn** [comment]
FWIW, I still haven't seen a strong argument for why Nulls.jl needs to be in Base. As it stands, I'd like to keep Base as minimal as possible, so the language can iterate independent of these datascience features. Is the issue simply not wanting to type `using Nulls`?

**Comment by quinnj** [comment]
@rofinn, technically, it's mainly about being able to have the syntax `T?` desugar to `Union{T, Null}`, as well as having a few definitions for `any`, `all`, and `==` on AbstractArrays that work correctly on `Union{T, Null}` eltype arrays, which could otherwise be considered type piracy, (though admittedly a lightweight form, so more like type-petty-theft). 

Another benefit I just thought of is the "proximity to the type system"; the `Null` type in Nulls.jl carries interesting semantics and use-cases with it as a form of "sentinel value for any type when part of a Union". Having the type definition "close" to where changes to the type system happen is nice because it gets more actively considered, much like @JeffBezanson's comment the other day [here](https://github.com/JuliaLang/julia/pull/23940#issuecomment-333439331).

I standby my arguments from above though that it also sends out a strong message that we've finally settled on a sensible standard that everyone can rely on.

**Comment by StefanKarpinski** [comment]
There are a few other reasons one might want a built in `null` type. Making `null || true == true` and `null || false == null` etc. work is one. There's also some consideration about whether `nothing` is appropriate as a missing or unknown value since it's more like C's `void` return type – which cannot be used as a value than like C's `null` value.

**Comment by rofinn** [comment]
> though admittedly a lightweight form, so more like type-petty-theft

Yeah, I'd be okay with that.

I'm still concerned that we're conflating missing values in software development and statistics. I can understand wanting something like `Union{Some{T}, Null}` and `Union{T, Null}` in Base instead of `Nullable`, but I feel like that's a separate conversation from how `null`s should be used for statistics. IMHO, I'd still expect `2 + null` to error and `2 + NA == NA`.

**Comment by richardreeve** [comment]
I'm not sure why you'd expect that. Adding a number to a missing value isn't an error, we just don't know what the answer is, surely? Or am I getting confused about what you're saying?

**Comment by andyferris** [comment]
> The problem with `~` was not that it was overloadable, but that it was a macro – that was just such a weird quirk of the operator that it ended up being regrettable almost as soon as we introduced it.

Yes, there were many problems here... however the specific problem I was referring to was scarcity - that it was *unique*. It was the *only* macro operator, and it becomes a race to see who claims it first. I decided not to use it specifically because I could foresee that another popular package might be used in tandem with whatever it was I was playing with at the time (sorry, the details elude me right now). It seems to me that if there were different "nulls" using `?` from different packages, its easy to foresee some user wanting (or being forced into by dependencies) to use more than one such package simultaneously.

**Comment by andyferris** [comment]
> I would recommend that the name of the missing value type should suggest its semantics. The word "null" is lousy since it suggests no semantics at all. There are at least three kinds of missing data: ...

+100 to this.

Not intentionally trying to muddy the waters, but I did start playing with `Optional{T}` being a container with a single `Union{T, Unknown{T}}` inside. I assumed it would be easier for users to know how to overload a method to take `Unknown{T}` - consider the projection of the function from all members of `T` and write down the most specific output (such as `true | Unknown{Bool}() == true` but `true & Unknown{Bool}() == Unknown{Bool}()`). There were issues discussed relating to covariance vs invariance and the pro's and con's of having the `T` parameter, but I still think "unknown" conveys the semantics for what data scientists want for missing data (well, at least in my case).

As for `null` / `Null`.... I do wonder whether this should be something that has to do with missing references and null pointers. (E.g. a `Nullable{T}` could be much like a `RefValue{T}` but be allowed to contain the zero pointer - together with the realization that `mutable struct T` is behaving more-or-less like a `Ref{T}` with some `getfield`/`setfield!` overloading, it might work out OK for uninitialized arrays of references, etc - I dunno, pretty speculative).

**Comment by rofinn** [comment]
@richardreeve Adding a number to a missing value could be an error depending on what the context of that missing value is (e.g., we don't know what `T` is in `Union{T, Null}`). Given the broad use of `null` (or `Nullable`) in software development I'd be inclined to err on the side of not propagating. I'm fine if there is a separate NA type which does propagate, but that provides context on what the missing value represents and how it is intended to be used.

**Comment by richardreeve** [comment]
@rofinn Okay, I **was** getting confused about what you were saying - I thought you were discussing using `null` in statistics from the context of the message... I agree that other contexts have different requirements. Hence your observation that we're conflating missing values in different situations I guess! I'd be very happy to see different names (e.g. `NA`) for the different contexts to avoid confusion, but that doesn't mean we don't know how each should work in its own right.

**Comment by andyferris** [comment]
At the moment I see these 4. @rofinn, @richardreeve - it seems to me that you are discussing 3 vs 4.

1. `nothing`, `Void` - the result of no code, e.g. `nothing === (begin; end)` (a part of the core language)
2. `null`, `isnull`, `Nullable` - something to do null pointers. Could use "undef" as the noun here as that's what users currently "see". (a part of the core language)
3. `Maybe{T}` - a container container zero or one elements of type `T`, useful for "software". Also, `Some{T}` vs. exceptional result types (error code, etc) would be a decent alternative here, or we currently have `Nullable{T}` for this. In both cases, requires manual unwrapping. (could be `Base`, a stdlib, or a package)
4. `Unknown` / `Unknown{T}` - semantically means that we expect in reality there exists some value of type `T` that should be here, but we don't currently know what it is. For data scientists, meant to be used unwrapped like `true |
 unknown == true`. (could be `Base`, a stdlib, or a package, with the caveat that perhaps we might decide we want this to interact with `if` in some way)

While I appreciate that other languages have lamented the sheer number of different types of "nullable" types, I see here four semantically distinct things here - it would be hard to accidentally use one of the 4 in place of another and not get errors relatively quickly.

**Comment by richardreeve** [comment]
@andyferris And to be honest, although I understand the importance of the others, **I only care about 4**. And though there may be technical reasons for putting each of them in Base or not, I think all of them should go in partly for "marketing" reasons to make a "strong statement" about support, and partly because as we've just perfectly demonstrated **this is very confusing** so we have to have a completely clear core definition of each so there is no disagreement about which one is being talked about and used at any given time!

**Comment by andyferris** [comment]
I very much appreciate that sentiment, @richardreeve. If I might offer some interpretation here for you - I think to many (most) here, the v1.0 milestone marks the beginning of some stability in the language - that we won't be changing the syntax or making too many breaking changes to the types in `Base` and so-forth (though I'm sure change will remain much more rapid than many other languages). It **definitely doesn't** mean the language supports all the features we want it to include.

Thus, the prerequisite to putting something completely new like in `Base` at this stage is that it has (nearly) universal support (which is a simple proxy/indication that no-one can see that we need to make breaking changes in the immediate future). I still feel we have a problem in reaching such universal consensus (and I dunno, the fact that brainstorming of radical changes is still occurring,might indicate we haven't really reached a point of stability/maturity, either).

The nice thing is that we can still have definite plans to add case 4 to `Base` or a standard library in v1.1, and release v1.0 sooner, which might benefit a wider range of users than delaying v1.0 for a long time or for making mistakes which means we have to release v2.0 soon with a lot of breaking changes. We can still make this "strong statement" at a future date. OTOH it's my opinion that we can (and should) tolerate a small delay to v1.0 to sort this stuff out, if that's at all feasible.

**Comment by mlhetland** [comment]
@andyferris Which other uses than no. 3 do you see for null pointers (i.e., no. 2)? And how is this different from the use of `nothing`, which seems to be the current plan for canonical handling of this? If code either returns something or not, you get `Union{T, Void}`, no? Assigning that to a variable seems to be an example of a case where there might or might not be a value. As opposed to the unknown value case (`NA`). And these are the main two variants that have been discussed, no? Something vs nothing and known vs unknown?

**Comment by JeffBezanson** [comment]
> `null || false == null`

I do think there are cases in Base where it will be nice/ergonomic to handle null correctly out of the box, but I really, really don't want to do this. The problem is that since you don't know whether `null` is true or false, you don't know whether the branch should be taken. This would also require putting Null not just in Base, but deep inside the compiler.

**Comment by nalimilan** [comment]
Yes, `null | false === null` is probably enough. `||` and `&&` are control flow operators, so I think they should behave like `if` and throw an error when they encounter `null`. For control flow you have to either execute a block or not execute it: null propagation is not really possible, you have to choose to treat `null` either as `true` or as `false`. Incidentally, this is was Erik Meijer and Eric Lippert [advised](https://github.com/JuliaLang/julia/pull/19034#issuecomment-265931467) based on the C# behavior and experience, while advocating at the same time three-valued logic for `&` and `|`.

**Comment by mlhetland** [comment]
It may be a cheat, but … it may not be entirely unreasonable to just forbid basing control flow on it, as you don’t know what you’re supposed to do? I mean, if you have `null || func()`, and `func` has side effects, returning `null` won’t cut it. Seems to me this should just be an error, just like using `null` as the condition of an `if` statement?

**Comment by mlhetland** [comment]
Eh. Right. Wrote my response in parallel with @nalimilan, there.

**Comment by StefanKarpinski** [comment]
> The problem is that since you don't know whether null is true or false, you don't know whether the branch should be taken.

Both. For `||` you evaluate until you hit a `true`, for `&&` you evaluate until you hit a `false`. I was not advocating for allowing `null` in other boolean contexts like `if` or `? :` – only in short-circuit chains. This behavior is not magical or crazy – it falls straight out of the 3VL truth tables and stopping evaluation as soon as the result of a chain is determined, which is exactly the rule we already use.

**Comment by mlhetland** [comment]
Sure, not magical or crazy :-) But it does seem that the short-circuit operators are idiomatically used as conditionals in Julia, in which case the lines seem to blur, perhaps? Things like `something || error("…")` are everywhere. In practice, then, `null` is masquerading as `false`. Might be fine, but if it doesn't make sense for `of` or `? :`, then maybe this doesn't make sense either?

More destructive in cases like `x == 0 && println("It's zero!")`, which would print if `x` is `null`, no? At the very least the documentation should warn against the use of such idioms whenever `null` might appear, I guess?

**Comment by nalimilan** [comment]
That makes sense, though that's low priority compared with other features we would want to work with `null`. If we add support for `null` in Base that question would certainly have to be considered.

**Comment by nalimilan** [comment]
At last!
